### PR TITLE
For #12544: Allow screen to turn off after media is paused/stopped

### LINKED ISSUE #12544: Allow screen to turn off after media is stopped
Issue seen on Fenix - https://github.com/mozilla-mobile/fenix/issues/25041

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21121)

### PR DESCRIPTION
**Pull request by indurs**
This fix allows the device to sleep when the media being played is paused/stopped after the system screen timeout is passed

https://user-images.githubusercontent.com/12825812/181486103-d49039a1-39a9-412b-b9a0-a44118d7cd47.mp4

Co-Authored-By: Mugurell <Mugurell@users.noreply.github.com>

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
Fixes #12544

### CODE REVIEW

File: components/feature/media/src/test/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeatureTest.kt [hunk 1]
```diff
@@ -320,4 +324,50 @@ class MediaSessionFullscreenFeatureTest {
 
         assertNotEquals(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE, externalActivity.requestedOrientation)
     }
+    @Suppress("Deprecation")
```
Comment (mcarare) [inline components/feature/media/src/test/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeatureTest.kt:327]: Just a drive-by question: What is the deprecation suppression for?

Comment (indurs) [reply components/feature/media/src/test/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeatureTest.kt:327]: Thank you for pointing that out !  It's not necessary for this test case. I copied a part of my previous test case and forgot to remove the deprecation line. Will update the code.

**Comment by rocketsroger** [comment]
Do we need to remove [this](https://github.com/mozilla-mobile/android-components/blob/164fb0a75b0ad7ab1f18bbb5e4aeb0d8c0db7eee/components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/Activity.kt#L33)?

**Comment by indurs** [comment]
You are right. We need to remove the flags from Activity extension methods related to immersive mode. The video in this PR was actually taken after the flags have been removed from the Activity extension methods but didn't commit the code changes for the Activity and ActivityTest
Checked with Mugurell to find out whether the flag is just for media https://github.com/mozilla-mobile/android-components/issues/12544. From the discussion I thought that we should commit the changes to MediaSessionFullscreenFeature in this PR and then commit the changes to Activity in a new PR. Probably I misunderstood. Shall we commit the changes to Activity and ActivityTest in this PR itself? @rocketsroger , @Mugurell  please let me know. Thank You

**Comment by Mugurell** [comment]
@indurs  In my comment from https://github.com/mozilla-mobile/android-components/issues/12544#issuecomment-1197838652 I was referring to having both changes 
- the new `processDeviceSleepMode` method and 
- refactoring `setAsImmersive`  & `exitImmersiveModeIfNeeded` 

in the same PR for them to land together. But maybe in separate commits to have better separation of intent.

Otherwise we will be left with a bug - not exiting immersive mode because of the `FLAG_KEEP_SCREEN_ON` check in `exitImmersiveModeIfNeeded`, flag which may have been removed by `processDeviceSleepMode` earlier.

So we need to also remove the responsability of un/setting the `FLAG_KEEP_SCREEN_ON` flag from the immersive mode extension methods and also the check for the flag such that `exitImmersiveModeIfNeeded` will become just `exitImmersiveMode`.

**Comment by indurs** [comment]
@Mugurell , Got it. Thank You ! Will update the code for the Activity as a separate commit in the same PR.

**Comment by Mugurell** [comment]
> @Mugurell , Got it. Thank You ! Will update the code for the Activity as a separate commit in the same PR.

Thank you for pushing this forward, it will be a great feature improvement!

**Comment by indurs** [comment]
@Mugurell , seems like we need to update BaseBrowserFragment in Fenix, if we refactor the exitImmersiveModeIfNeeded method to exitImmersiveMode. Do we need to create a PR in Fenix for this update?

**Comment by Mugurell** [comment]
Because of the naming change? Yes, we will need a ticket there also and a patch ready to be merged when the breaking change from AC gets released.
You can open a new ticket for this on Fenix or I can help also.

**Comment by indurs** [comment]
@Mugurell , could you please create a ticket in Fenix? I'll update the BaseBrowserFragment.

**Comment by Mugurell** [comment]
> @Mugurell , could you please create a ticket in Fenix? I'll update the BaseBrowserFragment.

Created https://github.com/mozilla-mobile/fenix/issues/26236

**Comment by indurs** [comment]
@Mugurell , the test case in  WebAppActivityFeatureTest is checking for FLAG_KEEP_SCREEN_ON since WebAppActivityFeature calls the enterToImmersiveMode method. Will remove the checking for flags in WebAppActivityFeatureTest

**Review by Mugurell — APPROVED** [review]
LGTM.
Waiting to have these changes accounted for in Fenix and Focus before merging
- [x] [#7437](https://github.com/mozilla-mobile/focus-android/pull/7437) PR on Focus
- [x] [#26238](https://github.com/mozilla-mobile/fenix/pull/26238) PR on Fenix

**Comment by Mugurell** [comment]
Both the patches from Fenix and Focus are ready to fix the breaking change this patch introduces so this looks ready to land.
